### PR TITLE
[Enhancement Request] FeedFlash 2.0 – A Scheduled News Summarization System

### PR DESCRIPTION
**Pull request by Arihant-Bhandari**
### Requesting to submit a pull request to the PyVerse repository.

---

#### Issue Title

FeedFlash – A Scheduled News Summarization System

- [x] I have provided the issue title.

---

#### Info about the Related Issue

FeedFlash is an automated news summarization system designed to run entirely on free-tier infrastructure. It continuously fetches live English-language headlines using NewsAPI every 4 hours (from 10 AM to 10 PM), extracts full article content with newspaper3k, and generates concise, easy-to-read summaries (~90 words) using a custom fine-tuned Flan-T5-Base model trained with PEFT/LoRA. Each article is processed with a 600-token input limit (covering about 400–800 words) and a 300-token output limit for the summary. Summaries and logs are stored as JSON files for seamless integration with the web app.

The user interface is built with Gradio, offering a modern, mobile-friendly experience where users can browse headlines and instantly view summaries—automatically scrolling to the summary section for convenience. The system is robustly automated via a custom GitHub Actions scheduler, ensuring updates every 4 hours and pushing new summaries to the repository and Hugging Face Space. FeedFlash is designed for easy reproducibility and maintenance, with all dependencies specified in requirements.txt, comprehensive documentation, and compatibility across local machines, Kaggle, GitHub Actions, and Hugging Face Spaces.

* Automated, scheduled news summarization and web delivery

* Fine-tuned Flan-T5-Base (LoRA/PEFT), 600→300 token summarization

* Gradio UI, mobile-friendly, error-handled, and fully reproducible

- [x] I have described the aim of the project.

---

#### Name

Arihant Bhandari

- [x] I have provided my name.

---

#### GitHub ID

GitHub ID: [Arihant-Bhandari](https://github.com/Arihant-Bhandari)

- [x] I have provided my GitHub ID.

---

#### Email ID

[arihant.bhandari.2004@gmail.com](arihant.bhandari.2004@gmail.com)

- [x] I have provided my email ID.

---

#### Identify Yourself

**SSOC 2025**

- [x] I have mentioned my participant role.

---

#### Closes
  
Closes: #1422 [Enhancement Request] FeedFlash – A Scheduled News Summarization System

- [x] I have provided the issue number.

---

#### Describe the Add-ons or Changes You've Made

Following changes were made:

## 1. `app/summarizer.py`

- **Model Used:**  
  - Loads a Flan-T5-Base model (`Arihant-Bhandari/feedflash-flan-t5`) that was fine-tuned using PEFT/LoRA, quantized to 4-bit during training for efficiency, and then dequantized back to full precision for CPU-based inference.
- **Lazy Loading:**  
  - Model and tokenizer are loaded only when needed, reducing memory and startup overhead.
- **Summarization Logic:**  
  - Accepts input text, applies tokenization and truncation, and generates a summary using beam search (num_beams=4).
  - Handles empty or invalid input with clear error messages.
- **Error Handling:**  
  - Catches and reports errors during model loading and summarization.

---

## 2. `backend/news_fetcher.py`

- **News Fetching:**  
  - Uses NewsAPI to fetch live English-language news headlines.
  - Fetches 20 international and 10 Indian articles per run, with no genre restrictions.
- **Article Extraction:**  
  - Extracts full article text using newspaper3k.
  - Filters out articles that are too short (<150 words) or too long (>2500 words).
- **Summarization Pipeline:**  
  - Summarizes each article using the dequantized Flan-T5-Base model via `summarize_text`.
  - Prints progress every 5 articles collected and summarized.
- **Output & Logging:**  
  - Saves summarized articles to `app/summaries.json`.
  - Stores detailed logs (including word counts and summary lengths) in `backend/fetch_log.json`, both in UTF-8 encoding.
- **Robustness:**  
  - Handles API errors, extraction failures, and summarization exceptions gracefully.

---

## 3. `app/app.py`

- **Data Loading:**  
  - Loads summarized articles from `app/summaries.json` with robust error handling for missing or malformed files.
- **Gradio Web UI:**  
  - Presents a modern, mobile-friendly interface for browsing headlines and summaries.
  - Headlines are displayed as radio buttons; selecting one updates the summary and source link.
- **Auto-Scroll Feature:**  
  - Automatically scrolls to the summary section when a new headline is selected, improving mobile usability.
- **Error Handling:**  
  - Displays user-friendly messages if no articles are available or if data is invalid.
- **UI Structure:**  
  - Clean separation of headline selector and summary display, with clear labeling and accessibility features.

---

## 4. `.github/workflows/fetch_news.yml` (Scheduled News Fetcher)

- **Custom Scheduler:**  
  - Runs at five key times daily (morning, pre-lunch, afternoon, evening, night wrap-up) using cron expressions.
- **Environment Setup:**  
  - Checks out the repository and sets up Python 3.10.
  - Installs all dependencies from `requirements.txt`.
- **Automated Pipeline:**  
  - Runs the news fetcher script (`python -m backend.news_fetcher`) with the required API key.
- **Sync to Hugging Face Space:**  
  - Clones the Hugging Face Space repo and copies updated `summaries.json` and `fetch_log.json` files.
  - Commits and pushes changes only if there are actual updates, avoiding redundant builds.
  - Handles missing files gracefully and logs each sync step for transparency.
- **Secure Operations:**  
  - Uses GitHub and Hugging Face secrets for authentication and API access.

---

## 5. **Model Upgrade (Global Across Files)**

- **Model:**  
  - Switched to a Flan-T5-Base model fine-tuned with PEFT/LoRA, quantized to 4-bit for efficient training and then dequantized for CPU/ONNX inference.
- **Token Limits:**  
  - Supports 600 input tokens (for longer news articles) and 300 output tokens (for richer summaries).
- **Training & Inference:**  
  - Model was trained on Kaggle with a P100 GPU using 4-bit quantization for speed and memory efficiency, then dequantized to full precision for fast, accurate CPU inference.

---

## Summary Table

| File/Component             | Key Features & Functionalities                                                                                 |
|----------------------------|---------------------------------------------------------------------------------------------------------------|
| `app/summarizer.py`        | Lazy-loaded summarizer, Flan-T5-Base (LoRA/PEFT, 4-bit quantized/dequantized), 600→300 tokens, robust errors |
| `backend/news_fetcher.py`  | Scheduled fetching, full article extraction, summarization, progress logging, output as JSON                  |
| `app/app.py`               | Gradio UI, mobile-friendly, auto-scroll to summary, robust data/error handling                                |
| `fetch_news.yml`           | Multi-time daily scheduler, dependency install, fetch/summarize pipeline, HF sync                             |
| **Model**                  | Flan-T5-Base, LoRA/PEFT, 4-bit quantized and dequantized, 600 input / 300 output tokens, trained on Kaggle    |

---

**These changes collectively deliver an automated, scalable, and user-friendly news summarization system, with robust scheduling, modern UI, and a powerful custom-trained summarization model that leverages 4-bit quantization for efficient training and dequantization for fast CPU inference.**

- [x] I have described my changes.

---

#### Type of Change
**Select the type of change:**  
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Code style update (formatting, local variables)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

---

#### How Has This Been Tested?

Model during training was monitored over entropy loss in terms of generation:

The training loss dropped from 2.36 (at step 100) to 1.95 (at step 14,517) over the course of training.

- [x] I have described my testing process.

---

#### Checklist
**Please confirm the following:**  
- [x] My code follows the guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly wherever it was hard to understand.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added things that prove my fix is effective or that my feature works.
- [x] Any dependent changes have been merged and published in downstream modules.

## Summary by CodeRabbit

* **New Features**
  * Introduced a fine-tuning pipeline for news summarization using a parameter-efficient, quantized FLAN-T5-base model.
  * Enhanced the news fetching process with category-based article collection (international and Indian news), improved filtering, and robust retry logic.
  * Automated news fetching and Hugging Face syncing now run multiple times daily with improved logging and error handling.
  * Added smooth scrolling in the web UI for easier navigation when selecting headlines.

* **Improvements**
  * Summarization now supports longer input and output texts for richer summaries.
  * Summaries and logs are stored in structured JSON files for better traceability.
  * Documentation updated to reflect the advanced, automated workflow, new features, and planned enhancements.
  * Expanded package requirements to support new model training and data processing features.

* **Bug Fixes**
  * Improved error handling and headline extraction in the web UI for greater stability.

* **Chores**
  * Workflow and codebase annotated and reorganized for clarity, with enhanced logging and safer file operations.

### CODE REVIEW

File: Deep_Learning/FeedFlash/model/feedflash-model-2.0.ipynb [hunk 1]
```diff
@@ -0,0 +1 @@
...:"<IPython.core.display.HTML object>","text/html":"\n    <div>\n      \n      <progress value='34' max='14517' style='width:300px; height:20px; vertical-align: middle;'></progress>\n      [   34/14517 00:52 < 6:34:30, 0.61 it/s, Epoch 0.01/3]\n    </div>\n    <table border=\"1\" class=\"dataframe\">\n  <thead>\n <tr style=\"text-align: left;\">\n      <th>Step</th>\n      <th>Training Loss</th>\n    </tr>\n  </thead>\n  <tbody>\n  </tbody>\n</table><p>"},"metadata":{}}],"execution_count":null},{"cell_type":"code","source":"merged_model = model.merge_and_unload()","metadata":{"trusted":true},"outputs":[],"execution_count":null},{"cell_type":"code","source":"merged_model.save_pretrained(\"feedflash-flan-t5-base\")\ntokenizer.save_pretrained(\"feedflash-flan-t5-base\")","metadata":{"trusted":true},"outputs":[],"execution_count":null},{"cell_type":"code","source":"from huggingface_hub import login\nlogin()","metadata":{"trusted":true},"outputs":[],"execution_count":null},{"cell_type":"code","source":"model.push_to_hub(\"Arihant-Bhandari/feedflash-flan-t5\")\ntokenizer.push_to_hub(\"Arihant-Bhandari/feedflash-flan-t5\")","metadata":{"trusted":true},"outputs":[],"execution_count":null}]}
```
[the remark this answers is not in the document]
Comment (Arihant-Bhandari) [reply Deep_Learning/FeedFlash/model/feedflash-model-2.0.ipynb:1]: your comment on using fp16 for model training would have been valid if the model was the standard full precision fp32 variant; the model that was trained in this file is a PEFT-LoRA based 4-bit quantized Flan T5 Base, trained on free-tier GPU access to a P100 on Kaggle Notebook.

As far as the comment on token size is concerned, the issue indeed mentions 1024-348 tokens format and trials on training a full scale flan t5 large was made that lead to unsuccessful runs, mainly due to the large size and huge training time for 50000 samples (98+ hours). By downgrading and using PEFT-LoRA configuration with 600-300 tokens framework we not only effectively cut down on time and space but with 4-bit quantization we make the model more lighter and faster cutting time down to 6 hours.

[the remark this answers is not in the document]
Comment (Arihant-Bhandari) [reply Deep_Learning/FeedFlash/model/feedflash-model-2.0.ipynb:1]: check latest news_fetcher.py and fetch_news.yml files please. they have been updated as per your comments.

File: Deep_Learning/FeedFlash/backend/news_fetcher.py [hunk 2]
```diff
@@ -11,102 +11,128 @@
...
+                if collected >= count:
+                    break
+                url = art.get("url")
+                if not url:
+                    continue
+                full_text = extract_full_article(url)
+                if not full_text:
+                    continue
+                word_count = len(full_text.split())
+                if word_count < 150 or word_count > 2500:
```
[the remark this answers is not in the document]
Comment (Arihant-Bhandari) [reply Deep_Learning/FeedFlash/backend/news_fetcher.py:76]: your suggestion to limit word count to 150 to 450 words is very stringent and 600 tokens handle majority of large context in content worth equivalent to 2500 words.

[the remark this answers is not in the document]
Comment (Arihant-Bhandari) [reply Deep_Learning/FeedFlash/backend/news_fetcher.py:76]: please let me know based on entire chain till now how many nitpicks were resolved and how many remain that need to be worked on. thank you.

[the remark this answers is not in the document]
Comment (Arihant-Bhandari) [reply Deep_Learning/FeedFlash/backend/news_fetcher.py]: let me know if there are any other nitpicks left that need to be solved.

[the remark this answers is not in the document]
Comment (Arihant-Bhandari) [reply Deep_Learning/FeedFlash/backend/news_fetcher.py]: please make appropriate changes in your original comment as well as project summary to include cotnent on chunking as well.

**Comment by kanishgit09** [comment]
@Arihant-Bhandari Brother, tell me in which folder you found this project data. please in detail. In PyVerse repo, there are many files he listed.

**Comment by Arihant-Bhandari** [comment]
> @Arihant-Bhandari Brother, tell me in which folder you found this project data. please in detail. In PyVerse repo, there are many files he listed.

Hi @kanishgit09 This is my original project that I contributed to PyVerse under the deep_learning/FeedFlash/ directory. You can find the dataset details in the project's README.md file.

This is my personal contribution to the repo. The PyVerse version has all the complete, working code that anyone can copy and run. Some parts are commented out as needed for the repo submission. I'm managing the live, deployed version of this project on my end.

If you want to work on this project or add new features, I'd love to collaborate! Feel free to contact me directly so we can coordinate. If your ideas are good, I'd be happy to work together on adding them since I handle the automated deployment.
